src/libsync/propagatebundle.cpp

```diff
@@ -0,0 +1,601 @@
+            start();
+        }
+    } else {
+        //_itemsToChecksum is already empty, send what is left in _itemsToSync.
+        startBundle();
+    }
+}
+
+void PropagateBundle::startBundle()
+{
+    if (_propagator->_abortRequested.fetchAndAddRelaxed(0)) {
+        return;
+    }
+
+    // job takes ownership of device via a QScopedPointer. Job deletes itself when finishing
+    MultipartJob* job = new MultipartJob(_propagator->account(), _propagator->account()->davFilesPath(), new QHttpMultiPart(QHttpMultiPart::RelatedType));
+
+    // start constructing the bundle metadata xml file
+    QByteArray xml = "<?xml version='1.0' encoding='UTF-8'?>\n"
+            "<d:multipart xmlns:d=\"DAV:\">\n";
+    int contentID = 0;
+
+    while (!_itemsToSync.isEmpty()){
+        const SyncFileItemPtr &item = _itemsToSync.takeFirst();
+        const QString fullFilePath = _propagator->getFilePath(item->_file);
+
+        if (!FileSystem::fileExists(fullFilePath)) {
+           itemDone(item, SyncFileItem::SoftError, tr("File Removed"));
+           continue;
+        }
+
+        time_t prevModtime = item->_modtime; // the _item value was set in PropagateUploadFileQNAM::start()
+        // but a potential checksum calculation could have taken some time during which the file could
+        // have been changed again, so better check again here.
+
+        item->_modtime = FileSystem::getModTime(fullFilePath);
+        if( prevModtime != item->_modtime ) {
+            _propagator->_anotherSyncNeeded = true;
+            itemDone(item, SyncFileItem::SoftError, tr("Local file changed during syncing. It will be resumed."));
+            continue;
+        }
+
+        item->_size = FileSystem::getSize(fullFilePath);
+
+        // But skip the file if the mtime is too close to 'now'!
+        // That usually indicates a file that is still being changed
+        // or not yet fully copied to the destination.
+        if (fileIsStillChanging(*item)) {
+            _propagator->_anotherSyncNeeded = true;
+            itemDone(item, SyncFileItem::SoftError, tr("Local file changed during sync."));
+            continue;
+        }
+
+        //TODO use Upload Device to support bandwith limitation on the client
+        QFile *file = new QFile(fullFilePath);
+        if (! file->open(QIODevice::ReadOnly)) {
+            qDebug() << "ERR: Could not prepare upload device: " << file->errorString();
+
+            // If the file is currently locked, we want to retry the sync
+            // when it becomes available again.
+            if (FileSystem::isFileLocked(fullFilePath)) {
+                emit _propagator->seenLockedFile(fullFilePath);
+            }
+
+            // Soft error because this is likely caused by the user modifying his files while syncing
+            itemDone(item, SyncFileItem::SoftError, tr("ERR: Could not prepare upload device"));
+        } else {
+            // add xml part for this specific file
+            xml += "  <d:part>\n"
```

Comment:
The important thing I wanted to get here, so it will be amortized O(1), so all fine with QString.